src/Features/Core/Portable/CodeStyle2/ExampleCodeStyleProvider.cs

```diff
@@ -0,0 +1,106 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.CodeStyle2
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    class ExampleDiagnosticAnalyzer : ExampleCodeStyleProvider.DiagnosticAnalyzer
+    {
+    }
+
+    [ExportCodeRefactoringProvider(LanguageNames.CSharp, Name = "Example code refactoring provider")]
+    class ExampleCodeRefactoringProvider : ExampleCodeStyleProvider.CodeRefactoringProvider
+    {
+    }
+
+    [ExportCodeFixProvider(LanguageNames.CSharp, Name = "Example code fix provider")]
+    class ExampleCodeFixProvider : ExampleCodeStyleProvider.CodeFixProvider
+    {
+    }
+
+    class ExampleCodeStyleProvider : AbstractCodeStyle2Provider<ExpressionBodyPreference, int, ExampleCodeStyleProvider>
+    {
+        public ExampleCodeStyleProvider()
+            : base(SomeOption,
+                   IDEDiagnosticIds.SomeID,
+                   SomeTitle,
+                   SomeMessage)
+        {
+        }
+
+        #region analysis
+
+        protected override bool DiagnosticsForOpenFileOnly(Workspace workspace)
+            => false;
+
+        protected override DiagnosticAnalyzerCategory GetDiagnosticAnalyzerCategory()
+            => DiagnosticAnalyzerCategory.SemanticSpanAnalysis;
+
+        protected override (ImmutableArray<int>, Action<SyntaxNodeAnalysisContext, ExpressionBodyPreference>) GetSyntaxNodeAction()
+        {
+            return (ImmutableArray.Create(1, 2, 3), AnalyzeSyntax);
+        }
+
+        private void AnalyzeSyntax(SyntaxNodeAnalysisContext context, ExpressionBodyPreference preference)
+        {
+            throw new NotImplementedException();
+        }
```

Comment:
subclass would implement this. but it would only actually be called if appropriate. subclass would not have to actually have to have the logic about "should i be running... or should only the refactoring be running?" :)